### PR TITLE
copy/cut select image and text together.

### PR DESCRIPTION
**Pull request by TheHeBoy**
copy/cut select image and text together.

https://github.com/singerdmx/flutter-quill/assets/49644098/75508d5f-e8ea-42e2-b9cb-087592515502

### CODE REVIEW

**Review by singerdmx — CHANGES_REQUESTED** [review]
Please format code
And update changelog

**Comment by singerdmx** [comment]
warning • Unused import: '../models/documents/style.dart' • lib/src/widgets/editor.dart:16:8 • unused_import
warning • Unused import: '../models/documents/style.dart' • lib/src/widgets/raw_editor.dart:23:8 • unused_import
   info • The import of '../../models/documents/document.dart' is unnecessary because all of the used elements are also provided by the import of '../../../flutter_quill.dart' • lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:7:8 • unnecessary_import
   info • The import of '../../models/documents/nodes/leaf.dart' is unnecessary because all of the used elements are also provided by the import of '../../../flutter_quill.dart' • lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:8:8 • unnecessary_import
   info • The import of '../editor.dart' is unnecessary because all of the used elements are also provided by the import of '../../../flutter_quill.dart' • lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:[10](https://github.com/singerdmx/flutter-quill/actions/runs/5560278692/jobs/10157148635?pr=1307#step:7:11):8 • unnecessary_import

**Comment by singerdmx** [comment]
Please don't use` ../../../flutter_quill.dart`

File: lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart [hunk 1]
```diff
@@ -3,13 +3,17 @@ import 'dart:math' as math;
 import 'package:flutter/rendering.dart';
 import 'package:flutter/widgets.dart';
 
+import '../../../flutter_quill.dart';
```
Comment (singerdmx) [inline lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:6]: Please don't use `import '../../../flutter_quill.dart';`

File: lib/src/models/documents/document.dart [hunk 2]
```diff
@@ -165,6 +165,12 @@ class Document {
     return (res.node as Line).collectAllIndividualStyles(res.offset, len);
   }
 
+  List<OffsetValue> collectAllIndividualStyleAndEmbed(int index, int len) {
```
Comment (singerdmx) [inline lib/src/models/documents/document.dart:168]: Can you add comments for this method?

**Review by singerdmx — CHANGES_REQUESTED** [review]

File: lib/src/widgets/controller.dart [hunk 3]
```diff
@@ -166,6 +166,12 @@ class QuillController extends ChangeNotifier {
     return styles;
   }
 
+  List<OffsetValue> getAllIndividualSelectionStylesAndEmbed() {
```
Comment (singerdmx) [inline lib/src/widgets/controller.dart:169]: Can you add comments for this meth or?

**Review by singerdmx — CHANGES_REQUESTED** [review]

File: lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart [hunk 4]
```diff
@@ -3,13 +3,17 @@ import 'dart:math' as math;
...
 import '../../models/documents/document.dart';
 import '../../models/documents/nodes/leaf.dart';
 import '../../utils/delta.dart';
 import '../editor.dart';
 
 mixin RawEditorStateSelectionDelegateMixin on EditorState
     implements TextSelectionDelegate {
+
+  bool isContainEmbed = false;
```
Comment (singerdmx) [inline lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:15]: This seems hacky

**Review by singerdmx — CHANGES_REQUESTED** [review]

File: lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart [hunk 5]
```diff
@@ -37,15 +42,29 @@ mixin RawEditorStateSelectionDelegateMixin on EditorState
...
+        } else {
+          widget.controller.formatTextStyle(
+              pos + offset,
+              i == pasteStyleAndEmbed.length - 1
+                  ? pastePlainText.length - offset
+                  : pasteStyleAndEmbed[i + 1].offset,
+              styleAndEmbed);
+        }
+      }
+    }else if(isContainEmbed){
```
Comment (singerdmx) [inline lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:60]: Can you remove the use of `isContainEmbed`?
For the least you can iterate through pasteStyleAndEmbed and check if there is any `style is Embeddabl`

Comment (TheHeBoy) [reply lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:60]:
```
  void _applyPasteStyle(String insertedText, int start) {
    if (insertedText == pastePlainText && pastePlainText != '') {
```
When I only paste images or backspace line head to the previous row, _applyPasteStyle method will be invoked and pastePlainText both are ''. I do not know whether paste the image. And that pasteStyleAndEmbed will not be clear for repetitive paste.

**Comment by TheHeBoy** [comment]
> Please format code And update changelog

What's mean by "update changelog"? Need me to add text  #[7.2.12]?

**Comment by singerdmx** [comment]
> > Please format code And update changelog
> 
> What's mean by "update changelog"? Need me to add text #[7.2.12]?

Yes

File: lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart (hunk 5 above)
Comment (singerdmx) [reply lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:60]: both `_adjustInsertedText` and `_applyPasteStyle` are invoked inside `set textEditingValue(TextEditingValue value)`

I am sure you have a better way

Comment (singerdmx) [reply lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:60]: I can schedule a call with to go through the logic

Comment (TheHeBoy) [reply lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:60]: The adjusted code is as follows, is ok?
```dart
  set textEditingValue(TextEditingValue value) {
    final cursorPosition = value.selection.extentOffset;
    final oldText = widget.controller.document.toPlainText();
    final newText = value.text;
    final diff = getDiff(oldText, newText, cursorPosition);
    if (diff.deleted == '' && diff.inserted == '') {
      // Only changing selection range
      widget.controller.updateSelection(value.selection, ChangeSource.LOCAL);
      return;
    }

    var isContainEmbed = false;
    var insertedText = diff.inserted;
    if (insertedText.codeUnits.contains(Embed.kObjectReplacementInt)) {
      isContainEmbed = true;
      insertedText = _adjustInsertedText(diff.inserted);
    }

    widget.controller.replaceText(
        diff.start, diff.deleted.length, insertedText, value.selection);

    _applyPasteStyle(insertedText, diff.start, isContainEmbed);
  }

  void _applyPasteStyle(String insertedText, int start, bool isContainEmbed) {
    if (insertedText == pastePlainText && pastePlainText != '') {
      final pos = start;
      for (var i = 0; i < pasteStyleAndEmbed.length; i++) {
        final offset = pasteStyleAndEmbed[i].offset;
        final styleAndEmbed = pasteStyleAndEmbed[i].value;

        if (styleAndEmbed is Embeddable) {
          widget.controller.replaceText(pos + offset, 0, styleAndEmbed, null);
        } else {
          widget.controller.formatTextStyle(
              pos + offset,
              i == pasteStyleAndEmbed.length - 1
                  ? pastePlainText.length - offset
                  : pasteStyleAndEmbed[i + 1].offset,
              styleAndEmbed);
        }
      }
    } else if (isContainEmbed) {
      final pos = start;
      for (var i = 0; i < pasteStyleAndEmbed.length; i++) {
        final offset = pasteStyleAndEmbed[i].offset;
        final style = pasteStyleAndEmbed[i].value;
        if (style is Embeddable) {
          widget.controller.replaceText(pos + offset, 0, style, null);
        }
      }
    }
  }

  String _adjustInsertedText(String text) {
    final sb = StringBuffer();
    for (var i = 0; i < text.length; i++) {
      if (text.codeUnitAt(i) == Embed.kObjectReplacementInt) {
        continue;
      }
      sb.write(text[i]);
    }
    return sb.toString();
  }
```

Comment (singerdmx) [reply lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:60]: Thanks! much better.
But `_applyPasteStyle` seems to have duplicated code which can be improved.
Also we may want to rename it to `_applyPasteStyleAndEmbed` ?

Comment (singerdmx) [reply lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:60]:
```
var containsEmbed = insertedText.codeUnits.contains(Embed.kObjectReplacementInt);
var insertedText = diff.inserted;
if (containsEmbed) {
      insertedText = _adjustInsertedText(diff.inserted);
}
```

Comment (TheHeBoy) [reply lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:60]: Yes, you are right.

Comment (singerdmx) [reply lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:60]:
```
void _applyPasteStyleAndEmbed(String insertedText, int start, bool containsEmbed) {
  if (insertedText == pastePlainText && pastePlainText != '' || containsEmbed) {
    final pos = start;
    for (var i = 0; i < pasteStyleAndEmbed.length; i++) {
      final offset = pasteStyleAndEmbed[i].offset;
      final styleAndEmbed = pasteStyleAndEmbed[i].value;

      if (styleAndEmbed is Embeddable) {
        widget.controller.replaceText(pos + offset, 0, styleAndEmbed, null);
      } else {
        widget.controller.formatTextStyle(
          pos + offset,
          i == pasteStyleAndEmbed.length - 1
              ? pastePlainText.length - offset
              : pasteStyleAndEmbed[i + 1].offset,
          styleAndEmbed);
      }
    }
  }
}

```

Comment (singerdmx) [reply lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:60]:
```
var containsEmbed = insertedText.codeUnits.contains(Embed.kObjectReplacementInt);
var insertedText = containsEmbed ? _adjustInsertedText(diff.inserted) : diff.inserted;
```

Comment (TheHeBoy) [reply lib/src/widgets/raw_editor/raw_editor_state_selection_delegate_mixin.dart:60]: Thanks! You make code more Concise

File: lib/src/models/documents/nodes/line.dart [hunk 6]
```diff
@@ -430,6 +430,44 @@ class Line extends Container<Leaf?> {
     return result;
   }
 
+  List<OffsetValue> collectAllIndividualStylesAndEmbed(int offset, int len,
```
Comment (singerdmx) [inline lib/src/models/documents/nodes/line.dart:433]:
```
List<OffsetValue> collectAllIndividualStylesAndEmbed(int offset, int len,
    {int beg = 0}) {
  final local = math.min(length - offset, len);
  final result = <OffsetValue>[];

  final data = queryChild(offset, true);
  var node = data.node as Leaf?;
  var pos = 0;

  while (node != null && pos < local) {
    if (node is Text || node.value is Embeddable) {
      pos = node.length - data.offset;
      result.add(OffsetValue(beg, node is Text ? node.style : node.value as Embeddable));
      pos += node.length;
    }
    if (!node.isLast) {
      node = node.next as Leaf;
    } else {
      break;
    }
  }

  final remaining = len - local;
  if (remaining > 0 && nextLine != null) {
    final rest = nextLine!.collectAllIndividualStylesAndEmbed(0, remaining, beg: local);
    result.addAll(rest);
  }

  return result;
}

```

Comment (singerdmx) [reply lib/src/models/documents/nodes/line.dart:433]: Please remove `collectAllIndividualStyles` method

File: lib/src/models/documents/document.dart (hunk 2 above)
Comment (singerdmx) [reply lib/src/models/documents/document.dart:168]: Please remove `collectAllIndividualStyles`

**Comment by singerdmx** [comment]
Need to replace the test:


  ```
  test('getAllIndividualSelectionStyles', () {
      controller.formatText(0, 2, Attribute.bold);
      final result = controller.getAllIndividualSelectionStyles();
      expect(result.length, 1);
      expect(result[0].offset, 0);
      expect(result[0].value, Style().put(Attribute.bold));
    });
```

File: lib/src/models/documents/nodes/line.dart (hunk 6 above)
Comment (TheHeBoy) [reply lib/src/models/documents/nodes/line.dart:433]: The above code has a worry mistake. Can't together paste text and image.And I haven't more much idea for this,The adjusted code is as follows:
```
  List<OffsetValue> collectAllIndividualStylesAndEmbed(int offset, int len,
      {int beg = 0}) {
    final local = math.min(length - offset, len);
    final result = <OffsetValue>[];

    final data = queryChild(offset, true);
    var node = data.node as Leaf?;
    if (node != null) {
      var pos = 0;
      if (node is Text || node.value is Embeddable) {
        pos = node.length - data.offset;
        result.add(OffsetValue(
            beg, node is Text ? node.style : node.value as Embeddable));
      }
      while (!node!.isLast && pos < local) {
        node = node.next as Leaf;
        if (node is Text || node.value is Embeddable) {
          result.add(OffsetValue(
              pos + beg, node is Text ? node.style : node.value as Embeddable));
          pos += node.length;
        }
      }
    }

    final remaining = len - local;
    if (remaining > 0 && nextLine != null) {
      final rest = nextLine!
          .collectAllIndividualStylesAndEmbed(0, remaining, beg: local);
      result.addAll(rest);
    }

    return result;
  }
```

**Comment by singerdmx** [comment]
Can you update this PR please?

**Pull request merged**